p2p/src/main/kotlin/maru/p2p/MaruPeerManager.kt

```diff
@@ -108,29 +121,70 @@ class MaruPeerManager(
   }
 
   override fun onConnect(peer: Peer) {
-    // TODO: here we could check if we want to be connected to that peer
-    if (p2pNetwork.peerCount > maxPeers) {
+    if (connectedPeers.size >= maxPeers) {
       // TODO: We could disconnect another peer here, based on some criteria
       peer.disconnectCleanly(DisconnectReason.TOO_MANY_PEERS)
+      return
     }
     val maruPeer = maruPeerFactory.createMaruPeer(peer)
-    connectedPeers.put(peer.id, maruPeer)
-    if (maruPeer.connectionInitiatedLocally()) {
-      maruPeer.sendStatus()
-    } else {
-      maruPeer.scheduleDisconnectIfStatusNotReceived(p2pConfig.statusUpdate.timeout)
-    }
+    statusExchangingMaruPeers[peer.id] = maruPeer
+    maruPeer
+      .awaitInitialStatus()
+      .thenApply { status ->
+        if (!status.forkIdHash.contentEquals(forkidHashProvider.currentForkIdHash())) {
+          log.debug(
+            "Peer={} has a different forkIdHash={} than expected={}. Disconnecting.",
+            peer.id,
+            status.forkIdHash,
+            forkidHashProvider.currentForkIdHash(),
+          )
+          maruPeer.disconnectCleanly(DisconnectReason.IRRELEVANT_NETWORK)
+        } else if (syncStatusProvider.getCLSyncStatus() == CLSyncStatus.SYNCING &&
+          status.latestBlockNumber < syncStatusProvider.getSyncTarget()!! - blocksBehindThreshold
+        ) {
+          log.debug(
+            "Peer={} is too far behind our target block number={} (peer's block number={}). Disconnecting.",
+            peer.id,
+            syncStatusProvider.getSyncTarget(),
+            status.latestBlockNumber,
+          )
+          maruPeer.disconnectCleanly(DisconnectReason.TOO_MANY_PEERS) // there is no better reason available
+        } else if (syncStatusProvider.getCLSyncStatus() == CLSyncStatus.SYNCED &&
+          status.latestBlockNumber < beaconChain.getLatestBeaconState().latestBeaconBlockHeader.number &&
+          tooManyPeersBehind(beaconChain.getLatestBeaconState().latestBeaconBlockHeader.number)
+        ) {
+          log.debug("Peer={} is behind and we already have too many peers behind. Disconnecting.", peer.id)
+          maruPeer.disconnectCleanly(DisconnectReason.TOO_MANY_PEERS) // there is no better reason available
+        } else {
+          connectedPeers.put(peer.id, maruPeer)
+          log.trace("Connected to peer={} with status={}", peer.id, status)
+        }
+      }.always {
+        statusExchangingMaruPeers.remove(peer.id)
+      }
   }
 
+  private fun tooManyPeersBehind(currentHead: ULong): Boolean =
+    connectedPeers.values.count { peer ->
+      peer.getStatus()!!.latestBlockNumber < currentHead - blocksBehindThreshold
+    } >= maxPeers / 10
```

Comment:
What's the idea behind this? If we have more than 10% of peers behinds behind head we connect another node also behind head?

Reply:
This is only used when we are in sync, so we would be very close to head.